include/envoy/server/resource_monitor.h

```diff
@@ -18,39 +21,103 @@ struct ResourceUsage {
+  virtual ~ResourceUpdateCallbacks() = default;
 
   /**
-   * Notifies caller of updated resource usage.
+   * Called when the request for updated resource usage succeeds.
+   * @param usage the updated resource usage
    */
-  class Callbacks {
-  public:
-    virtual ~Callbacks() = default;
-
-    /**
-     * Called when the request for updated resource usage succeeds.
-     * @param usage the updated resource usage
-     */
-    virtual void onSuccess(const ResourceUsage& usage) PURE;
-
-    /**
-     * Called when the request for updated resource usage fails.
-     * @param error the exception caught when trying to get updated resource usage
-     */
-    virtual void onFailure(const EnvoyException& error) PURE;
-  };
+  virtual void onSuccess(const ResourceUsage& usage) PURE;
+
+  /**
+   * Called when the request for updated resource usage fails.
+   * @param error the exception caught when trying to get updated resource usage
+   */
+  virtual void onFailure(const EnvoyException& error) PURE;
+};
+
+class ReactiveResourceUpdateCallbacks {
+public:
+  virtual ~ReactiveResourceUpdateCallbacks() = default;
+
+  /**
+   * Called when the request for updated resource usage succeeds.
+   * @param usage the updated resource usage
+   */
+  virtual void onSuccess(const uint64_t usage) PURE;
+
+  /**
+   * Called when the request for updated resource usage fails.
+   * todo may need propagate error
+   */
+  virtual void onFailure() PURE;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+};
+
+class ResourceMonitor {
+public:
+  virtual ~ResourceMonitor() = default;
 
   /**
    * Recalculate resource usage.
    * This must be non-blocking so if RPCs need to be made they should be
    * done asynchronously and invoke the callback when finished.
    */
-  virtual void updateResourceUsage(Callbacks& callbacks) PURE;
+  virtual void updateResourceUsage(ResourceUpdateCallbacks& callbacks) PURE;
+};
+
+class ReactiveResourceMonitor {
+public:
+  ReactiveResourceMonitor() = default;
+  virtual ~ReactiveResourceMonitor() = default;
+  virtual bool tryAllocateResource(uint64_t increment) PURE;
+  virtual bool tryDeallocateResource(uint64_t decrement) PURE;
+  virtual uint64_t currentResourceUsage() const PURE;
+  virtual uint64_t maxResourceUsage() const PURE;
 };
 
 using ResourceMonitorPtr = std::unique_ptr<ResourceMonitor>;
 
+using ReactiveResourceMonitorPtr = std::unique_ptr<ReactiveResourceMonitor>;
+
+// Example of reactive resource monitor. To be removed.
+class ActiveConnectionsResourceMonitor : public ReactiveResourceMonitor {
+public:
+  ActiveConnectionsResourceMonitor(uint64_t max_active_conns)
+      : max_(max_active_conns), current_(0){};
+
+  bool tryAllocateResource(uint64_t increment) {
+    uint64_t new_val = (current_ += increment);
+    if (new_val >= max_) {
+      current_ -= increment;
+      return false;
+    }
+    return true;
+  }
+
+  bool tryDeallocateResource(uint64_t decrement) {
```

Comment:
I'm not sure what caller code can/should if deallocation fails.  If deallocation ever fails we have a serious logic error.

I guess an alternate API could involve a movable resource tracking object that releases the resource on destruction.  That approach could reduce the risk of mismatched calls to allocate/deallocate the resource.  That said, the resource tracking object approach may be overkill given that we expect this API to be used in a few places, one example being allocate just before creation of a server connection and de-allocation on server connection destruction.